File: legacy-python/src/websearch_mcp/fetch/resolver.py
"""Fast paths for sources a coding agent hits constantly and that have a cleaner-than-HTML
representation: GitHub (raw files, READMEs, issues, gists), PyPI, npm, StackOverflow, arXiv,
and llms.txt. Each returns a `Fetched` with kind markdown/text, or None to fall through."""

from __future__ import annotations

import logging
import os
import re
from urllib.parse import urlsplit

import httpx

from .extract import html_snippet_to_markdown
from .transport import Fetched

log = logging.getLogger(__name__)

CODE_EXT = {
    "py": "python", "js": "javascript", "ts": "typescript", "tsx": "tsx", "jsx": "jsx", "rs": "rust",
    "go": "go", "java": "java", "kt": "kotlin", "swift": "swift", "rb": "ruby", "php": "php",
    "c": "c", "h": "c", "cpp": "cpp", "cc": "cpp", "hpp": "cpp", "cs": "csharp", "sh": "bash",
    "bash": "bash", "zsh": "bash", "yml": "yaml", "yaml": "yaml", "toml": "toml", "json": "json",
    "xml": "xml", "html": "html", "css": "css", "scss": "scss", "sql": "sql", "proto": "protobuf",
    "tf": "hcl", "dockerfile": "dockerfile", "lua": "lua", "ex": "elixir", "exs": "elixir",
    "scala": "scala", "hs": "haskell", "ml": "ocaml", "clj": "clojure", "dart": "dart", "r": "r",
    "jl": "julia", "zig": "zig", "nim": "nim", "vue": "vue", "svelte": "svelte", "ini": "ini",
    "cfg": "ini", "mk": "makefile", "makefile": "makefile", "gradle": "groovy", "ps1": "powershell",
}
TEXT_EXT = {"md", "mdx", "markdown", "rst", "txt", "adoc", "org"}


def _gh_headers(accept: str) -> dict[str, str]:
    h = {"Accept": accept, "User-Agent": "websearch-mcp/0.1 (+https://github.com)"}
    tok = os.environ.get("GITHUB_TOKEN")
    if tok:
        h["Authorization"] = f"Bearer {tok}"
    return h


def wrap_file(name: str, text: str) -> tuple[str, str]:
    """Return (kind, body): markdown-ish text verbatim, code fenced with a language."""
    ext = name.rsplit(".", 1)[-1].lower() if "." in name else name.lower()
    if ext in TEXT_EXT:
        return "markdown", text
    lang = CODE_EXT.get(ext, "")
    fence = "```"
    while fence in text:
        fence += "`"
    return "markdown", f"{fence}{lang}\n{text.rstrip()}\n{fence}\n"


async def _get(client: httpx.AsyncClient, url: str, headers: dict[str, str] | None = None) -> httpx.Response | None:
    try:
        r = await client.get(url, headers=headers, follow_redirects=True)
    except httpx.HTTPError as e:
        log.debug("fast path GET failed %s: %s", url, e)
        return None
    if r.status_code != 200:
        log.debug("fast path GET %s -> %s", url, r.status_code)
        return None
    return r


async def _github(url: str, parts, client: httpx.AsyncClient) -> Fetched | None:
    host = parts.netloc.lower()
    segs = [s for s in parts.path.split("/") if s]

    if host == "raw.githubusercontent.com" or host == "gist.githubusercontent.com":
        r = await _get(client, url)
        if r is None:
            return None
        kind, body = wrap_file(segs[-1] if segs else "", r.text)
        return Fetched(url, url, kind, body, "github-raw")

    if host == "gist.github.com" and len(segs) >= 2:
        r = await _get(client, f"https://api.github.com/gists/{segs[1]}", _gh_headers("application/vnd.github+json"))
        if r is None:
            return None
        data = r.json()
        out = [f"# Gist: {data.get('description') or segs[1]}"]
        for name, f in (data.get("files") or {}).items():
            content = f.get("content") or ""
            _, body = wrap_file(name, content)
            out.append(f"## {name}\n\n{body}")
        return Fetched(url, url, "markdown", "\n\n".join(out) + "\n", "github-gist")

    if host != "github.com" or len(segs) < 2:
        return None
    owner, repo = segs[0], segs[1]

    if len(segs) == 2:
        r = await _get(client, f"https://api.github.com/repos/{owner}/{repo}/readme", _gh_headers("application/vnd.github.raw+json"))
        if r is None:
            return None
        meta = await _get(client, f"https://api.github.com/repos/{owner}/{repo}", _gh_headers("application/vnd.github+json"))
        head = ""
        if meta is not None:
            m = meta.json()
            head = (
                f"# {m.get('full_name')}\n\n{m.get('description') or ''}\n\n"
                f"Stars: {m.get('stargazers_count')} · Language: {m.get('language')} · "
                f"Default branch: {m.get('default_branch')} · Updated: {str(m.get('pushed_at'))[:10]}\n\n---\n\n"
            )
        return Fetched(url, url, "markdown", head + r.text, "github-readme")

    if len(segs) >= 4 and segs[2] in {"blob", "raw"}:
        raw = f"https://raw.githubusercontent.com/{owner}/{repo}/{'/'.join(segs[3:])}"
        r = await _get(client, raw)
        if r is None:
            return None
        kind, body = wrap_file(segs[-1], r.text)
        return Fetched(url, url, kind, body, "github-raw")

    if len(segs) >= 4 and segs[2] in {"issues", "pull"} and segs[3].isdigit():
        num = segs[3]
        r = await _get(client, f"https://api.github.com/repos/{owner}/{repo}/issues/{num}", _gh_headers("application/vnd.github+json"))
        if r is None:
            return None
        issue = r.json()
        out = [
            f"# {issue.get('title')} (#{num})",
            (
                f"{owner}/{repo} · {issue.get('state')} · by {(issue.get('user') or {}).get('login')} · "
                f"{str(issue.get('created_at'))[:10]} · {issue.get('comments', 0)} comments"
            ),
            "",
            issue.get("body") or "(no description)",
        ]
        c = await _get(client, f"https://api.github.com/repos/{owner}/{repo}/issues/{num}/comments?per_page=15", _gh_headers("application/vnd.github+json"))
        if c is not None:
            for cm in c.json():
                out.append(f"\n---\n**{(cm.get('user') or {}).get('login')}** ({str(cm.get('created_at'))[:10]}):\n\n{cm.get('body') or ''}")
        return Fetched(url, url, "markdown", "\n".join(out) + "\n", "github-issue")

    return None


async def _pypi(parts, client: httpx.AsyncClient, url: str) -> Fetched | None:
    m = re.match(r"^/(?:project|pypi)/([^/]+)/?", parts.path)
    if not m:
        return None
    r = await _get(client, f"https://pypi.org/pypi/{m.group(1)}/json", {"Accept": "application/json"})
    if r is None:
        return None
    info = r.json().get("info") or {}
    urls = info.get("project_urls") or {}
    head = [
        f"# {info.get('name')} {info.get('version')}",
        info.get("summary") or "",
        "",
        f"Requires: {info.get('requires_python') or '?'} · License: {info.get('license') or '?'}",
        "Links: " + " · ".join(f"{k}: {v}" for k, v in list(urls.items())[:6]),
        "",
        "---",
        "",
    ]
    desc = info.get("description") or ""
    if (info.get("description_content_type") or "").startswith("text/x-rst"):
        desc = "(README is reStructuredText)\n\n" + desc
    return Fetched(url, url, "markdown", "\n".join(head) + desc + "\n", "pypi")


async def _npm(parts, client: httpx.AsyncClient, url: str) -> Fetched | None:
    m = re.match(r"^/package/((?:@[^/]+/)?[^/]+)", parts.path)
    if not m:
        return None
    r = await _get(client, f"https://registry.npmjs.org/{m.group(1)}", {"Accept": "application/json"})
    if r is None:
        return None
    data = r.json()
    latest = (data.get("dist-tags") or {}).get("latest")
    head = [
        f"# {data.get('name')} {latest or ''}",
        data.get("description") or "",
        "",
        f"Homepage: {data.get('homepage') or '?'} · Repo: {((data.get('repository') or {}).get('url') if isinstance(data.get('repository'), dict) else data.get('repository')) or '?'}",
        "",
        "---",
        "",
    ]
    return Fetched(url, url, "markdown", "\n".join(head) + (data.get("readme") or "(no README)") + "\n", "npm")


async def _stackoverflow(parts, client: httpx.AsyncClient, url: str) -> Fetched | None:
    m = re.match(r"^/(?:questions|q)/(\d+)", parts.path)
    if not m:
        return None
    qid = m.group(1)
    base = "https://api.stackexchange.com/2.3"
    q = await _get(client, f"{base}/questions/{qid}?site=stackoverflow&filter=withbody")
    if q is None:
        return None
    items = q.json().get("items") or []
    if not items:
        return None
    qi = items[0]
    out = [
        f"# {html_snippet_to_markdown(qi.get('title') or '').strip()}",
        f"Score {qi.get('score')} · {qi.get('answer_count')} answers · tags: {', '.join(qi.get('tags') or [])}",
        "",
        html_snippet_to_markdown(qi.get("body") or ""),
    ]
    a = await _get(client, f"{base}/questions/{qid}/answers?site=stackoverflow&order=desc&sort=votes&filter=withbody&pagesize=4")
    if a is not None:
        answers = a.json().get("items") or []
        answers.sort(key=lambda x: (not x.get("is_accepted"), -x.get("score", 0)))
        for ans in answers:
            tag = "Accepted answer" if ans.get("is_accepted") else "Answer"
            out.append(f"\n---\n## {tag} (score {ans.get('score')})\n\n{html_snippet_to_markdown(ans.get('body') or '')}")
    return Fetched(url, url, "markdown", "\n".join(out) + "\n", "stackoverflow")


def rewrite_url(url: str) -> str:
    """Cheap URL rewrites that yield a better representation of the same content."""
    parts = urlsplit(url)
    host = parts.netloc.lower()
    if host in {"arxiv.org", "www.arxiv.org"}:
        # Canonical abstract page; /pdf/ URLs are left alone (they go through the PDF path).
        m = re.match(r"^/abs/([\w.]+?)(?:v\d+)?/?$", parts.path)
        if m:
            return f"https://arxiv.org/abs/{m.group(1)}"
    return url


async def resolve_fast_path(url: str, client: httpx.AsyncClient) -> Fetched | None:
    parts = urlsplit(url)
    host = parts.netloc.lower()
    try:
        if host.endswith(("github.com", "githubusercontent.com")):
            return await _github(url, parts, client)
        if host in {"pypi.org", "www.pypi.org"}:
            return await _pypi(parts, client, url)
        if host in {"www.npmjs.com", "npmjs.com"}:
            return await _npm(parts, client, url)
        if host in {"stackoverflow.com", "www.stackoverflow.com"}:
            return await _stackoverflow(parts, client, url)
    except Exception as e:  # noqa: BLE001 — fast paths are best-effort
        log.warning("fast path failed for %s: %s", url, e)
    return None


_llms_cache: dict[str, str | None] = {}


async def llms_txt(url: str, client: httpx.AsyncClient) -> str | None:
    """Return the site's /llms.txt content if it exists (cached per origin)."""
    parts = urlsplit(url)
    origin = f"{parts.scheme}://{parts.netloc}"
    if origin in _llms_cache:
        return _llms_cache[origin]
    r = await _get(client, f"{origin}/llms.txt", {"Accept": "text/plain, text/markdown"})
    text = None
    if r is not None:
        ct = r.headers.get("content-type", "")
        body = r.text
        if "html" not in ct and "<html" not in body[:500].lower() and len(body.strip()) > 50:
            text = body
    _llms_cache[origin] = text
    return text

File: packages/core/src/fetch/resolver.ts
/**
 * Fast paths for sources a coding agent hits constantly that have a documented, cleaner-than-HTML
 * representation: GitHub (raw files, READMEs, issues, gists), PyPI, npm, StackOverflow, arXiv, llms.txt.
 * These use the sites' public APIs under their API terms. Each returns a `Fetched` or null to fall through.
 */

import type { Fetched, HttpLike } from "./types.js";
import { htmlSnippetToMarkdown } from "./extract.js";

const CODE_EXT: Record<string, string> = {
  py: "python",
  js: "javascript",
  mjs: "javascript",
  cjs: "javascript",
  ts: "typescript",
  mts: "typescript",
  tsx: "tsx",
  jsx: "jsx",
  rs: "rust",
  go: "go",
  java: "java",
  kt: "kotlin",
  swift: "swift",
  rb: "ruby",
  php: "php",
  c: "c",
  h: "c",
  cpp: "cpp",
  cc: "cpp",
  hpp: "cpp",
  cs: "csharp",
  sh: "bash",
  bash: "bash",
  zsh: "bash",
  yml: "yaml",
  yaml: "yaml",
  toml: "toml",
  json: "json",
  xml: "xml",
  html: "html",
  css: "css",
  scss: "scss",
  sql: "sql",
  proto: "protobuf",
  tf: "hcl",
  dockerfile: "dockerfile",
  lua: "lua",
  ex: "elixir",
  exs: "elixir",
  scala: "scala",
  hs: "haskell",
  ml: "ocaml",
  clj: "clojure",
  dart: "dart",
  r: "r",
  jl: "julia",
  zig: "zig",
  nim: "nim",
  vue: "vue",
  svelte: "svelte",
  ini: "ini",
  cfg: "ini",
  mk: "makefile",
  makefile: "makefile",
  gradle: "groovy",
  ps1: "powershell",
};
const TEXT_EXT = new Set(["md", "mdx", "markdown", "rst", "txt", "adoc", "org"]);

/**
 * Documented public API endpoints we use; robots.txt governs crawling, API terms govern these.
 * Hosts that serve both pages and an API (Wikipedia, MDN, crates.io, PyPI) are matched by path.
 */
export const API_ENDPOINTS: Array<[host: string, pathPrefix: string]> = [
  ["api.github.com", "/"],
  ["raw.githubusercontent.com", "/"],
  ["gist.githubusercontent.com", "/"],
  ["registry.npmjs.org", "/"],
  ["api.stackexchange.com", "/"],
  ["export.arxiv.org", "/"],
  ["pypi.org", "/pypi/"],
  ["crates.io", "/api/"],
  ["developer.mozilla.org", "/api/"],
  ["en.wikipedia.org", "/w/api.php"],
];

export function isApiUrl(url: string): boolean {
  try {
    const u = new URL(url);
    const host = u.hostname.toLowerCase();
    return API_ENDPOINTS.some(([h, p]) => host === h && u.pathname.startsWith(p));
  } catch {
    return false;
  }
}

function ghHeaders(accept: string): Record<string, string> {
  const h: Record<string, string> = { Accept: accept };
  const tok = process.env.GITHUB_TOKEN;
  if (tok) h.Authorization = `Bearer ${tok}`;
  return h;
}

/** Return (kind, body): markdown-ish text verbatim, code fenced with a language. */
export function wrapFile(name: string, text: string): ["markdown", string] {
  const ext = name.includes(".") ? name.split(".").pop()!.toLowerCase() : name.toLowerCase();
  if (TEXT_EXT.has(ext)) return ["markdown", text];
  const lang = CODE_EXT[ext] ?? "";
  let fence = "```";
  while (text.includes(fence)) fence += "`";
  return ["markdown", `${fence}${lang}\n${text.replace(/\s+$/, "")}\n${fence}\n`];
}

/** A fast-path result: markdown we produced ourselves from an API, attributed to `source`. */
function markdownResult(url: string, body: string, source: string): Fetched {
  return { url, finalUrl: url, kind: "markdown", body, source, status: 200, contentType: "text/markdown", headers: {} };
}

async function getOk(http: HttpLike, url: string, headers?: Record<string, string>) {
  try {
    const r = await http(url, { headers });
    return r.status === 200 ? r : null;
  } catch {
    return null;
  }
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type Json = Record<string, any>;

async function github(url: string, u: URL, http: HttpLike): Promise<Fetched | null> {
  const host = u.hostname.toLowerCase();
  const segs = u.pathname.split("/").filter(Boolean);

  if (host === "raw.githubusercontent.com" || host === "gist.githubusercontent.com") {
    const r = await getOk(http, url);
    if (!r) return null;
    const [, body] = wrapFile(segs[segs.length - 1] ?? "", await r.text());
    return markdownResult(url, body, "github-raw");
  }

  if (host === "gist.github.com" && segs.length >= 2) {
    const r = await getOk(http, `https://api.github.com/gists/${segs[1]}`, ghHeaders("application/vnd.github+json"));
    if (!r) return null;
    const data = (await r.json()) as Json;
    const out = [`# Gist: ${data.description || segs[1]}`];
    for (const [name, f] of Object.entries((data.files ?? {}) as Record<string, Json>)) {
      const [, body] = wrapFile(name, String(f.content ?? ""));
      out.push(`## ${name}\n\n${body}`);
    }
    return markdownResult(url, out.join("\n\n") + "\n", "github-gist");
  }

  if (host !== "github.com" || segs.length < 2) return null;
  const [owner, repo] = segs;

  if (segs.length === 2) {
    const r = await getOk(
      http,
      `https://api.github.com/repos/${owner}/${repo}/readme`,
      ghHeaders("application/vnd.github.raw+json"),
    );
    if (!r) return null;
    const meta = await getOk(
      http,
      `https://api.github.com/repos/${owner}/${repo}`,
      ghHeaders("application/vnd.github+json"),
    );
    let head = "";
    if (meta) {
      const m = (await meta.json()) as Json;
      head =
        `# ${m.full_name}\n\n${m.description ?? ""}\n\n` +
        `Stars: ${m.stargazers_count} · Language: ${m.language} · Default branch: ${m.default_branch} · Updated: ${String(m.pushed_at).slice(0, 10)}\n\n---\n\n`;
    }
    return markdownResult(url, head + (await r.text()), "github-readme");
  }

  if (segs.length >= 4 && (segs[2] === "blob" || segs[2] === "raw")) {
    const raw = `https://raw.githubusercontent.com/${owner}/${repo}/${segs.slice(3).join("/")}`;
    const r = await getOk(http, raw);
    if (!r) return null;
    const [, body] = wrapFile(segs[segs.length - 1], await r.text());
    return markdownResult(url, body, "github-raw");
  }

  if (segs.length >= 4 && segs[2] === "tree") {
    // github.com/o/r/tree/<ref>/<path...> — branch names may contain "/", so try ref boundaries in turn.
    const rest = segs.slice(3);
    for (let refLen = 1; refLen <= Math.min(rest.length, 3); refLen++) {
      const ref = rest.slice(0, refLen).join("/");
      const path = rest.slice(refLen).join("/");
      const r = await getOk(
        http,
        `https://api.github.com/repos/${owner}/${repo}/contents/${path}?ref=${encodeURIComponent(ref)}`,
        ghHeaders("application/vnd.github+json"),
      );
      if (!r) continue;
      const items = (await r.json()) as Json[] | Json;
      if (!Array.isArray(items)) continue; // a file, not a directory
      const dirs = items.filter((i) => i.type === "dir").sort((a, b) => String(a.name).localeCompare(String(b.name)));
      const files = items.filter((i) => i.type !== "dir").sort((a, b) => String(a.name).localeCompare(String(b.name)));
      const lines = [
        `# ${owner}/${repo} — ${path || "/"} @ ${ref}`,
        "",
        ...dirs.map((d) => `- ${d.name}/`),
        ...files.map((f) => `- ${f.name}${f.size ? ` (${f.size} B)` : ""}`),
        "",
        `${items.length} entries. Fetch a file with its blob URL, e.g. https://github.com/${owner}/${repo}/blob/${ref}/${path ? path + "/" : ""}<name>.`,
      ];
      return markdownResult(url, lines.join("\n") + "\n", "github-tree");
    }
    return null;
  }

  if (segs.length >= 3 && segs[2] === "releases") {
    const tag = segs[3] === "tag" ? segs[4] : undefined;
    const api = tag
      ? `https://api.github.com/repos/${owner}/${repo}/releases/tags/${encodeURIComponent(tag)}`
      : `https://api.github.com/repos/${owner}/${repo}/releases?per_page=10`;
    const r = await getOk(http, api, ghHeaders("application/vnd.github+json"));
    if (!r) return null;
    const data = (await r.json()) as Json | Json[];
    const rels = Array.isArray(data) ? data : [data];
    const out = [`# ${owner}/${repo} releases${tag ? ` — ${tag}` : " (latest 10)"}`, ""];
    for (const rel of rels) {
      out.push(
        `## ${rel.name || rel.tag_name} (${rel.tag_name}, ${String(rel.published_at ?? rel.created_at).slice(0, 10)}${rel.prerelease ? ", pre-release" : ""})`,
        "",
        String(rel.body ?? "(no notes)").trim(),
        "",
      );
    }
    return markdownResult(url, out.join("\n") + "\n", "github-releases");
  }

  if (segs.length >= 4 && (segs[2] === "issues" || segs[2] === "pull") && /^\d+$/.test(segs[3])) {
    const num = segs[3];
    const r = await getOk(
      http,
      `https://api.github.com/repos/${owner}/${repo}/issues/${num}`,
      ghHeaders("application/vnd.github+json"),
    );
    if (!r) return null;
    const issue = (await r.json()) as Json;
    const out = [
      `# ${issue.title} (#${num})`,
      `${owner}/${repo} · ${issue.state} · by ${issue.user?.login} · ${String(issue.created_at).slice(0, 10)} · ${issue.comments ?? 0} comments`,
      "",
      issue.body || "(no description)",
    ];
    const c = await getOk(
      http,
      `https://api.github.com/repos/${owner}/${repo}/issues/${num}/comments?per_page=15`,
      ghHeaders("application/vnd.github+json"),
    );
    if (c) {
      for (const cm of (await c.json()) as Json[]) {
        out.push(`\n---\n**${cm.user?.login}** (${String(cm.created_at).slice(0, 10)}):\n\n${cm.body ?? ""}`);
      }
    }
    return markdownResult(url, out.join("\n") + "\n", "github-issue");
  }
  return null;
}

async function pypi(url: string, u: URL, http: HttpLike): Promise<Fetched | null> {
  const m = /^\/(?:project|pypi)\/([^/]+)\/?/.exec(u.pathname);
  if (!m) return null;
  const r = await getOk(http, `https://pypi.org/pypi/${m[1]}/json`, { Accept: "application/json" });
  if (!r) return null;
  const info = (((await r.json()) as Json).info ?? {}) as Json;
  const urls = (info.project_urls ?? {}) as Record<string, string>;
  const head = [
    `# ${info.name} ${info.version}`,
    info.summary ?? "",
    "",
    `Requires: ${info.requires_python || "?"} · License: ${info.license || "?"}`,
    "Links: " +
      Object.entries(urls)
        .slice(0, 6)
        .map(([k, v]) => `${k}: ${v}`)
        .join(" · "),
    "",
    "---",
    "",
  ];
  let desc = String(info.description ?? "");
  if (String(info.description_content_type ?? "").startsWith("text/x-rst"))
    desc = "(README is reStructuredText)\n\n" + desc;
  return markdownResult(url, head.join("\n") + desc + "\n", "pypi");
}

async function npm(url: string, u: URL, http: HttpLike): Promise<Fetched | null> {
  const m = /^\/package\/((?:@[^/]+\/)?[^/]+)/.exec(u.pathname);
  if (!m) return null;
  const r = await getOk(http, `https://registry.npmjs.org/${m[1]}`, { Accept: "application/json" });
  if (!r) return null;
  const data = (await r.json()) as Json;
  const latest = data["dist-tags"]?.latest ?? "";
  const repo = typeof data.repository === "object" ? data.repository?.url : data.repository;
  const head = [
    `# ${data.name} ${latest}`,
    data.description ?? "",
    "",
    `Homepage: ${data.homepage || "?"} · Repo: ${repo || "?"}`,
    "",
    "---",
    "",
  ];
  return markdownResult(url, head.join("\n") + (data.readme || "(no README)") + "\n", "npm");
}

async function stackoverflow(url: string, u: URL, http: HttpLike): Promise<Fetched | null> {
  const m = /^\/(?:questions|q)\/(\d+)/.exec(u.pathname);
  if (!m) return null;
  const base = "https://api.stackexchange.com/2.3";
  const q = await getOk(http, `${base}/questions/${m[1]}?site=stackoverflow&filter=withbody`);
  if (!q) return null;
  const items = (((await q.json()) as Json).items ?? []) as Json[];
  if (!items.length) return null;
  const qi = items[0];
  const who = (o: Json | undefined) => (o?.display_name ? String(o.display_name) : "unknown");
  const out = [
    `# ${htmlSnippetToMarkdown(String(qi.title ?? "")).trim()}`,
    `Asked by ${who(qi.owner as Json)} · score ${qi.score} · ${qi.answer_count} answers · tags: ${(qi.tags ?? []).join(", ")}`,
    "",
    htmlSnippetToMarkdown(String(qi.body ?? "")),
  ];
  const a = await getOk(
    http,
    `${base}/questions/${m[1]}/answers?site=stackoverflow&order=desc&sort=votes&filter=withbody&pagesize=4`,
  );
  if (a) {
    const answers = (((await a.json()) as Json).items ?? []) as Json[];
    answers.sort((x, y) => Number(!!y.is_accepted) - Number(!!x.is_accepted) || (y.score ?? 0) - (x.score ?? 0));
    for (const ans of answers) {
      const tag = ans.is_accepted ? "Accepted answer" : "Answer";
      const link = `https://stackoverflow.com/a/${ans.answer_id}`;
      out.push(
        `\n---\n## ${tag} by ${who(ans.owner as Json)} (score ${ans.score}) — ${link}\n\n${htmlSnippetToMarkdown(String(ans.body ?? ""))}`,
      );
    }
  }
  out.push(
    "\n---\nContent from Stack Overflow is licensed CC BY-SA 4.0: attribute the authors and link to the posts when reusing it.",
  );
  return markdownResult(url, out.join("\n") + "\n", "stackoverflow");
}

/** Cheap URL rewrites that yield a better representation of the same content. */
export function rewriteUrl(url: string): string {
  const u = new URL(url);
  if (u.hostname === "arxiv.org" || u.hostname === "www.arxiv.org") {
    const m = /^\/abs\/([\w.]+?)(?:v\d+)?\/?$/.exec(u.pathname);
    if (m) return `https://arxiv.org/abs/${m[1]}`;
  }
  return url;
}

export async function resolveFastPath(url: string, http: HttpLike): Promise<Fetched | null> {
  const u = new URL(url);
  const host = u.hostname.toLowerCase();
  try {
    if (host.endsWith("github.com") || host.endsWith("githubusercontent.com")) return await github(url, u, http);
    if (host === "pypi.org" || host === "www.pypi.org") return await pypi(url, u, http);
    if (host === "www.npmjs.com" || host === "npmjs.com") return await npm(url, u, http);
    if (host === "stackoverflow.com" || host === "www.stackoverflow.com") return await stackoverflow(url, u, http);
  } catch {
    return null; // fast paths are best-effort
  }
  return null;
}

const llmsCache = new Map<string, { text: string | null; at: number }>();
const LLMS_TTL_MS = 60 * 60_000;

/** Return the site's /llms.txt content if it exists (cached per origin for an hour, not forever). */
export async function llmsTxt(url: string, http: HttpLike): Promise<string | null> {
  const u = new URL(url);
  const origin = `${u.protocol}//${u.host}`;
  const hit = llmsCache.get(origin);
  if (hit && Date.now() - hit.at < LLMS_TTL_MS) return hit.text;
  let text: string | null = null;
  const r = await getOk(http, `${origin}/llms.txt`, { Accept: "text/plain, text/markdown" });
  if (r) {
    const ct = r.headers["content-type"] ?? "";
    const body = await r.text();
    if (!ct.includes("html") && !/<html/i.test(body.slice(0, 500)) && body.trim().length > 50) text = body;
  }
  if (llmsCache.size > 500) llmsCache.clear();
  llmsCache.set(origin, { text, at: Date.now() });
  return text;
}
